File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/Expression.scala
package org.enso.compiler.core.ir

import org.enso.compiler.core.Implicits.{ShowPassData, ToStringHelper}
import org.enso.compiler.core.IR
import org.enso.compiler.core.Identifier
import org.enso.compiler.core.IR.{indentLevel, mkIndent, randomId}

import java.util.UUID
import scala.jdk.FunctionConverters.enrichAsScalaFromFunction

trait Expression extends IR {

  /** Performs a recursive traversal of the IR, potentially transforming it.
    *
    * @param fn the function to apply across the IR
    * @return the IR, potentially transformed
    */
  def transformExpressions(
    fn: PartialFunction[Expression, Expression]
  ): Expression = {
    if (fn.isDefinedAt(this)) {
      fn(this)
    } else {
      mapExpressions(_.transformExpressions(fn))
    }
  }

  /** @inheritdoc */
  override def mapExpressions(
    fn: java.util.function.Function[Expression, Expression]
  ): Expression

  /** @inheritdoc */
  override def setLocation(location: Option[IdentifiedLocation]): Expression

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): Expression
}

object Expression {

  // TODO Remove suspended blocks from Enso.

  /** A block expression.
    *
    * @param expressions the expressions in the block
    * @param returnValue the final expression in the block
    * @param location    the source location that the node corresponds to
    * @param suspended   whether or not the block is suspended
    * @param passData    the pass metadata associated with this node
    * @param diagnostics compiler diagnostics for this node
    */
  sealed case class Block(
    expressions: List[Expression],
    returnValue: Expression,
    location: Option[IdentifiedLocation],
    suspended: Boolean             = false,
    passData: MetadataStorage      = MetadataStorage(),
    diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends Expression
      with IRKind.Primitive {
    var id: UUID @Identifier = randomId

    /** Creates a copy of `this`.
      *
      * @param expressions the expressions in the block
      * @param returnValue the final expression in the block
      * @param location    the source location that the node corresponds to
      * @param suspended   whether or not the block is suspended
      * @param passData    the pass metadata associated with this node
      * @param diagnostics compiler diagnostics for this node
      * @param id          the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      expressions: List[Expression]        = expressions,
      returnValue: Expression              = returnValue,
      location: Option[IdentifiedLocation] = location,
      suspended: Boolean                   = suspended,
      passData: MetadataStorage            = passData,
      diagnostics: DiagnosticStorage       = diagnostics,
      id: UUID @Identifier                 = id
    ): Block = {
      val res = Block(
        expressions,
        returnValue,
        location,
        suspended,
        passData,
        diagnostics
      )
      res.id = id
      res
    }

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Block =
      copy(
        expressions = expressions.map(
          _.duplicate(
            keepLocations,
            keepMetadata,
            keepDiagnostics,
            keepIdentifiers
          )
        ),
        returnValue = returnValue.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        location = if (keepLocations) location else None,
        passData = if (keepMetadata) passData.duplicate else MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else randomId
      )

    /** @inheritdoc */
    override def setLocation(location: Option[IdentifiedLocation]): Block =
      copy(location = location)

    /** @inheritdoc */
    override def mapExpressions(
      fn: java.util.function.Function[Expression, Expression]
    ): Block = {
      copy(
        expressions = expressions.map(fn.asScala),
        returnValue = fn(returnValue)
      )
    }

    /** @inheritdoc */
    override def toString: String =
      s"""
         |Expression.Block(
         |expressions = $expressions,
         |returnValue = $returnValue,
         |location = $location,
         |suspended = $suspended,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] = expressions :+ returnValue

    /** @inheritdoc */
    override def showCode(indent: Int): String = {
      val newIndent = indent + indentLevel
      val expressionsStr = expressions
        .map(mkIndent(newIndent) + _.showCode(newIndent))
        .mkString("\n")
      val returnStr = mkIndent(newIndent) + returnValue.showCode(newIndent)

      s"\n$expressionsStr\n$returnStr"
    }
  }

  /** A binding expression of the form `name = expr`
    *
    * To create a binding that binds no available name, set the name of the
    * binding to an [[Name.Blank]] (e.g. _ = foo a b).
    *
    * @param name        the name being bound to
    * @param expression  the expression being bound to `name`
    * @param location    the source location that the node corresponds to
    * @param passData    the pass metadata associated with this node
    * @param diagnostics compiler diagnostics for this node
    */
  sealed case class Binding(
    name: Name,
    expression: Expression,
    location: Option[IdentifiedLocation],
    passData: MetadataStorage      = MetadataStorage(),
    diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends Expression
      with IRKind.Primitive {

    var id: UUID @Identifier = randomId

    /** Creates a copy of `this`.
      *
      * @param name        the name being bound to
      * @param expression  the expression being bound to `name`
      * @param location    the source location that the node corresponds to
      * @param passData    the pass metadata associated with this node
      * @param diagnostics compiler diagnostics for this node
      * @param id          the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      name: Name                           = name,
      expression: Expression               = expression,
      location: Option[IdentifiedLocation] = location,
      passData: MetadataStorage            = passData,
      diagnostics: DiagnosticStorage       = diagnostics,
      id: UUID @Identifier                 = id
    ): Binding = {
      val res = Binding(name, expression, location, passData, diagnostics)
      res.id = id
      res
    }

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Binding =
      copy(
        name = name.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        expression = expression.duplicate(
          keepLocations,
          keepMetadata,
          keepDiagnostics,
          keepIdentifiers
        ),
        location = if (keepLocations) location else None,
        passData = if (keepMetadata) passData.duplicate else MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else randomId
      )

    /** @inheritdoc */
    override def setLocation(location: Option[IdentifiedLocation]): Binding =
      copy(location = location)

    /** @inheritdoc */
    override def mapExpressions(
      fn: java.util.function.Function[Expression, Expression]
    ): Binding = {
      copy(name = name.mapExpressions(fn), expression = fn(expression))
    }

    /** @inheritdoc */
    override def toString: String =
      s"""
         |Expression.Binding(
         |name = $name,
         |expression = $expression,
         |location = $location
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] = List(name, expression)

    /** @inheritdoc */
    override def showCode(indent: Int): String =
      s"${name.showCode(indent)} = ${expression.showCode(indent)}"
  }
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/Expression.java
package org.enso.compiler.core.ir;

import java.util.function.Function;
import org.enso.compiler.core.IR;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Option;
import scala.PartialFunction;
import scala.collection.immutable.List;

public interface Expression extends IR {

  @Override
  Expression mapExpressions(Function<Expression, Expression> fn);

  /**
   * Performs a recursive traversal of the IR, potentially transforming it.
   *
   * @param fn the function to apply across the IR
   * @return the IR, potentially transformed
   */
  default Expression transformExpressions(PartialFunction<Expression, Expression> fn) {
    if (fn.isDefinedAt(this)) {
      return fn.apply(this);
    } else {
      return mapExpressions(e -> e.transformExpressions(fn));
    }
  }

  @Override
  Expression setLocation(Option<IdentifiedLocation> location);

  default Expression duplicate() {
    return duplicate(true, true, true, false);
  }

  @Override
  Expression duplicate(
      boolean keepLocations,
      boolean keepMetadata,
      boolean keepDiagnostics,
      boolean keepIdentifiers);

  @GenerateIR(interfaces = {Expression.class, IRKind.Primitive.class})
  final class Block extends ExpressionBlockGen {
    /**
     * @param expressions the expressions in the block
     * @param returnValue the final expression in the block
     * @param suspended whether or not the block is suspended
     */
    @GenerateFields
    public Block(
        @IRChild List<Expression> expressions,
        @IRChild Expression returnValue,
        @IRField boolean suspended,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData) {
      super(expressions, returnValue, suspended, identifiedLocation, passData);
    }

    public static Builder builder() {
      return new Builder().expressions(nil()).suspended(false);
    }

    public Builder copyBuilder() {
      return new Builder(this);
    }

    @SuppressWarnings("unchecked")
    private static <T> scala.collection.immutable.List<T> nil() {
      return (scala.collection.immutable.List<T>) scala.collection.immutable.Nil$.MODULE$;
    }

    @Override
    public String showCode(int indent) {
      var newIndent = indent + indentLevel;
      var expressionsStr =
          expressions().map(e -> IR.mkIndent(newIndent) + e.showCode(newIndent)).mkString("\n");
      var returnStr = IR.mkIndent(newIndent) + returnValue().showCode(newIndent);
      return "\n" + expressionsStr + "\n" + returnStr;
    }
  }

  /**
   * A binding expression of the form {@code name = expr}.
   *
   * <p>To create a binding that binds no available name, set the name of the binding to an {@link
   * Name.Blank} (e.g. _ = foo a b).
   */
  @GenerateIR(interfaces = {Expression.class, IRKind.Primitive.class})
  final class Binding extends ExpressionBindingGen {

    /**
     * @param name the name being bound to
     * @param expression the expression being bound to `name`
     */
    @GenerateFields
    public Binding(
        @IRChild Name name,
        @IRChild Expression expression,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData) {
      super(name, expression, identifiedLocation, passData);
    }

    public static Builder builder() {
      return new Builder();
    }

    public Builder copyBuilder() {
      return new Builder(this);
    }

    @Override
    public Binding mapExpressions(Function<Expression, Expression> fn) {
      var newName = name().mapExpressions(fn);
      var newExpr = fn.apply(expression());
      return new Builder(this).name(newName).expression(newExpr).build();
    }

    @Override
    public String showCode(int indent) {
      return name().showCode(indent) + " = " + expression().showCode(indent);
    }
  }
}
